From 066b52a669dbd5c06b9bb0d1f48a2052b78ec030 Mon Sep 17 00:00:00 2001
From: Demi Marie Obenour <demi@invisiblethingslab.com>
Date: Sat, 20 Mar 2021 14:30:03 -0400
Subject: [PATCH 10/17] Force GCC to behave as RPM expects

RPM assumes that integers and pointers have twos-complement
representation and semantics.  In particular, it assumes that it can
freely create out-of-bounds pointers.  It also assumes that strict
aliasing is not enforced.  Therefore, GCC needs to be made to behave
accordingly.

Also fix a case where a zero-sized tag data entry was wrongly accepted.
---
 configure.ac | 9 ++++++---
 lib/header.c | 8 ++++----
 2 files changed, 10 insertions(+), 7 deletions(-)

diff --git a/configure.ac b/configure.ac
index 34ea85f..40efb26 100644
--- a/configure.ac
+++ b/configure.ac
@@ -37,7 +37,7 @@ fi
 AS=${AS-as}
 AC_SUBST(AS)
 if test "$GCC" = yes; then
-    cflags_to_try="-fno-strict-aliasing -fstack-protector -Wempty-body"
+    cflags_to_try="-fno-strict-aliasing -fstack-protector -Wempty-body -fwrapv-pointer"
     AC_MSG_CHECKING([supported compiler flags])
     old_cflags=$CFLAGS
     echo
@@ -50,9 +50,12 @@ if test "$GCC" = yes; then
         ],[])
         CFLAGS=$old_cflags
     done
-    RPMCFLAGS="-fPIC -DPIC -D_REENTRANT -Wall -Wpointer-arith -Wmissing-prototypes -Wstrict-prototypes $RPMCFLAGS"
+    RPMCFLAGS="-fPIC -DPIC -D_REENTRANT -Wall -Wpointer-arith \
+        -Wmissing-prototypes -Wstrict-prototypes $RPMCFLAGS \
+        -fno-delete-null-pointer-checks -fwrapv \
+        -fno-strict-overflow -fno-strict-aliasing"
 fi
-AC_SUBST(RPMCFLAGS)
+AC_SUBST([RPMCFLAGS])
 
 AC_SYS_LARGEFILE
 
diff --git a/lib/header.c b/lib/header.c
index f5e62bb..b547096 100644
--- a/lib/header.c
+++ b/lib/header.c
@@ -312,9 +312,9 @@ static rpmRC hdrblobVerifyInfo(hdrblob blob, char **emsg, int exact_size)
 	/* Verify the data actually fits */
 	len = dataLength(info.type, ds + info.offset,
 			 info.count, 1, ds + blob->dl);
-	end = info.offset + len;
-	if (hdrchkRange(blob->dl, end) || len <= 0)
+	if (len <= 0 || blob->dl - info.offset < len)
 	    goto err;
+	end = info.offset + len;
     }
     return 0; /* Everything ok */
 
@@ -461,7 +461,7 @@ static int dataLength(rpm_tagtype_t type, rpm_constdata_t p, rpm_count_t count,
     const char * se = pend;
     int length = 0;
 
-    assert(!hdrchkType(type));
+    assert(!hdrchkType(type) && !hdrchkData(count));
 
     switch (type) {
     case RPM_STRING_TYPE:
@@ -490,7 +490,7 @@ static int dataLength(rpm_tagtype_t type, rpm_constdata_t p, rpm_count_t count,
 	if (typeSizes[type] == -1)
 	    return -1;
 	length = typeSizes[(type & 0xf)] * count;
-	if (length < 0 || (se && (s + length) > se))
+	if (length <= 0 || (se && length > se - s))
 	    return -1;
 	break;
     }
-- 
2.30.2

